plugin/healthchecker/checkers/icmp.go

```diff
@@ -0,0 +1,216 @@
+	}
+
+	return &ICMPChecker{
+		logger:       logger,
+		isPrivileged: prm.IsPrivileged,
+		timeout:      prm.Timeout,
+	}, nil
+}
+
+func (c ICMPChecker) Check(endpoint string) bool {
+	isV4 := isIPv4(endpoint)
+	ip := net.ParseIP(endpoint)
+	if ip == nil {
+		return false
+	}
+
+	prm, err := c.getConnParams(isV4)
+	if err != nil {
+		log.Debugf("failed to get icmp params: %s", err.Error())
+		return false
+	}
+
+	conn, err := icmp.ListenPacket(prm.Network, prm.ListenAddress)
+	if err != nil {
+		log.Debugf("listen icpm packet %s: %s", prm.Network, err.Error())
+		return false
+	}
+	defer conn.Close()
+
+	if err = c.writeMsg(conn, prm.Msg, ip); err != nil {
+		log.Debugf("write icmp msg: %s", err.Error())
+		return false
+	}
+
+	if err = c.readMsg(conn, prm); err != nil {
+		log.Debugf("read icmp msg: %s", err.Error())
+		return false
+	}
+
+	return true
+}
+
+func (c ICMPChecker) writeMsg(conn *icmp.PacketConn, msg []byte, ip net.IP) error {
+	if _, err := conn.WriteTo(msg, c.formAddress(ip)); err != nil {
+		return fmt.Errorf("write to '%s': %w", ip.String(), err)
+	}
+	return nil
+}
+
+func (c ICMPChecker) readMsg(conn *icmp.PacketConn, prm icmpParams) error {
+	rb := make([]byte, 1500)
+
+	if err := conn.SetReadDeadline(time.Now().Add(c.timeout)); err != nil {
+		return fmt.Errorf("set read deadline: %w", err)
+	}
+
+	n, peer, err := conn.ReadFrom(rb)
+	if err != nil {
+		return fmt.Errorf("read from: %w", err)
+	}
+
+	rm, err := icmp.ParseMessage(prm.Proto, rb[:n])
+	if err != nil {
+		return fmt.Errorf("parse msg: %w", err)
+	}
+
+	switch rm.Type {
+	case prm.ReplyType:
+		return nil
+	default:
+		return fmt.Errorf("got %+v from %v; want echo reply", rm, peer)
+	}
+}
+
+func (c ICMPChecker) formAddress(ip net.IP) net.Addr {
+	var addr net.Addr
+	if c.isPrivileged {
+		addr = &net.IPAddr{IP: ip}
+	} else {
+		addr = &net.UDPAddr{IP: ip}
+	}
+	return addr
+}
+
+func (c ICMPChecker) getConnParams(isV4 bool) (icmpParams, error) {
+	prm := icmpParams{}
+
+	msg := &icmp.Message{
+		Code: 0,
+		Body: &icmp.Echo{
+			ID: os.Getpid() & 0xffff, Seq: 1,
+			Data: []byte(""),
+		},
+	}
+
+	if isV4 {
+		msg.Type = ipv4.ICMPTypeEcho
+
+		prm.ListenAddress = "0.0.0.0"
+		prm.ReplyType = ipv4.ICMPTypeEchoReply
+		prm.Proto = 1 // https://godoc.org/golang.org/x/net/internal/iana ProtocolICMP
+
+		if c.isPrivileged {
+			prm.Network = "ip4:icmp"
+		} else {
+			prm.Network = "udp4"
+		}
+	} else {
+		msg.Type = ipv6.ICMPTypeEchoRequest
+
+		prm.ListenAddress = "::"
+		prm.ReplyType = ipv6.ICMPTypeEchoReply
+		prm.Proto = 58 // https://godoc.org/golang.org/x/net/internal/iana ProtocolIPv6ICMP
+
+		if c.isPrivileged {
+			prm.Network = "ip6:ipv6-icmp"
+		} else {
+			prm.Network = "udp6"
+		}
+	}
+
+	msgRaw, err := msg.Marshal(nil)
+	if err != nil {
+		return icmpParams{}, fmt.Errorf("marshal icmp msg: %w", err)
+	}
+
+	prm.Msg = msgRaw
+
+	return prm, nil
+}
+
+func isIPv4(str string) bool {
```

Comment:
Checking like `strings.Count(address, ":") < 2` look simpler, imo. But it's just a matter of taste